### PR TITLE
Update core data types 

### PR DESCRIPTION
**Pull request by Liamsi**
closes: #16 

## Description

This PR aims to align the current core data types, Block and Header, with what we have in the spec with the constraint that we agreed not to tackle [immediate execution](#3) immediately (no pun intended). Further, I've limited this to all changes that do not change *any* other parts besides that (e.g. I did not move evidence over to `Block.Data` as this will make this less contained and more difficult to review). Also I didn't change any hashing hashing logic s.t. none of the current tests break. Both these constraints will be addressed in separate smaller PRs.

### CODE REVIEW

File: types/block.go [hunk 1]
```diff
@@ -806,18 +829,53 @@ func (sh SignedHeader) StringIndented(indent string) string {
...
 	// NOTE: not all txs here are valid.  We're just agreeing on the order first.
 	// This means that block.AppHash does not include these txs.
 	Txs Txs `json:"txs"`
 
+	// Intermediate state roots of the Txs included in block.Height
+	// and executed by state state @ block.Height+1.
+	//
+	// TODO: replace with a dedicated type `IntermediateStateRoot`
+	// as soon as we settle on the format / sparse Merkle tree etc
+	IntermediateStateRoots []tmbytes.HexBytes
```
Comment (Liamsi) [inline types/block.go:855]: If you wonder about `tmbytes.HexBytes`: just think of it as `[]byte` (it's just for encoding really).

**Comment by Liamsi** [comment]
Note that CI currently doesn't run. I ran all tests locally. Will add using github actions instead of circleci soon. In the meantime I'd recommend enabling circleci for this org!

**Comment by Liamsi** [comment]
reopening and closing to trigger circleci

File: types/block.go [hunk 2]
```diff
@@ -806,18 +829,53 @@ func (sh SignedHeader) StringIndented(indent string) string {
...
+type Message struct {
+	// NamespaceID defines the namespace of this message, i.e. the
+	// namespace it will use in the namespaced Merkle tree.
+	//
+	// TODO: spec out constrains and
+	// introduce dedicated type instead of just []byte
+	NamespaceID []byte
+
+	// Data is the actual data contained in the message
+	// (e.g. a block of an optimistic rollup sidechain).
```
Comment (adlerjohn) [inline types/block.go:875]:
```suggestion
	// (e.g. a block of a virtual sidechain).
```

File: types/block.go [hunk 3]
```diff
@@ -34,13 +34,34 @@ const (
...
-	mtx        sync.Mutex
-	Header     `json:"header"`
-	Data       `json:"data"`
-	Evidence   EvidenceData `json:"evidence"`
-	LastCommit *Commit      `json:"last_commit"`
+	mtx sync.Mutex
+
+	Header                 `json:"header"`
+	Data                   `json:"data"`
+	DataAvailabilityHeader DataAvailabilityHeader
```
Comment (adlerjohn) [inline types/block.go:62]: Is there a reason there's no "json:..." for this line? Is this something that's deprecated anyways?

File: types/block.go [hunk 4]
```diff
@@ -34,13 +34,34 @@ const (
...
+// coded version of the data in Block.Data.
+// Therefor the original Block.Data is arranged in a
+// k × k matrix, which is then "extended" to a
+// 2k × 2k matrix applying multiple times Reed-Solomon encoding.
+// For details see Section 5.2: https://arxiv.org/abs/1809.09044
+type DataAvailabilityHeader struct {
+	// RowRoot_j 	= root((M_{j,1} || M_{j,2} || ... || M_{j,2k} ))
+	RowsRoots [][]byte
+	// ColumnRoot_j = root((M_{1,j} || M_{2,j} || ... || M_{2k,j} ))
+	ColumnRoots [][]byte
```
Comment (adlerjohn) [inline types/block.go:47]: If there's not a type for hash digests then we should add one. Is there is then `[]byte` can be replaced with it.

File: types/block.go (hunk 3 above)
Comment (Liamsi) [reply types/block.go:62]: Hmm, these annotations are (directly) used for JSON encoding. On the long run I'd like to separate core types that have methods and hold logic  from serialization concerns. But maybe we should keep these annotations. I think in geth there are also annoations but they are used to *generate* types that purely deal with serialization (better than what we have here). For now you are right, there should als be annotations. Will add them in a followup PR (please don't resolve this comment until then).

File: types/block.go (hunk 4 above)
Comment (Liamsi) [reply types/block.go:47]: There is none. At least a type alias would be good! Will add in a followup PR!

File: types/block.go (hunk 3 above)
Comment (Liamsi) [reply types/block.go]: NVM, added them in https://github.com/LazyLedger/lazyledger-core/pull/17/commits/6373e8850fb21107fc42d703e6509fb6cf131f3d.

**Pull request merged**